### PR TITLE
Use DenseMap for PtsMap.

### PR DESCRIPTION
**Pull request by mbarbar**
As title. Required specifying CondVar for DenseMapInfo. The hash could be improved, but it shouldn't be too bad.

Some Andersen's stats, in seconds:
mutt  39.4 -> 32.3
mruby 9.77 -> 6.194

Edit: Also, I decided to implement the empty and tombstone keys like LLVM does, with ~0 and ~1. It think with such a NodeID, no analysis can scale...
https://llvm.org/doxygen/DenseMapInfo_8h_source.html#118

### CODE REVIEW

**Comment by yuleisui** [comment]
One flow-sensitive test case failed.

https://travis-ci.org/github/SVF-tools/SVF/jobs/732846128#L1496

**Comment by mbarbar** [comment]
`-ander` results with...
`std::map`:
mruby: 19s
mutt: 49.8s

`std::unordered_map`:
mruby: 16.8s
mutt: 41.67s

Like we discussed, the STL maps are simpler and `DenseMap` is faster, thoughn that would need more tests to determine. Regardless, if we're moving to STL anyway, it doesn't matter. And like we discussed, perhaps someday a map tailored to SVF could be written.

Though I'm not 100% sure about the new `operator==` for `LocationSet` (required by `unordered_map`), perhaps @kisslune can take a look.

File: include/DDA/DDAVFSolver.h [hunk 1]
```diff
@@ -33,8 +33,8 @@ class DDAVFSolver
     typedef std::map<DPIm, CPtSet> DPImToCPtSetMap;
```
Comment (yuleisui) [inline include/DDA/DDAVFSolver.h:33]: can we also replace all the std::map with SVFMap?

**Review by yuleisui — COMMENTED** [review]
I can see there are still quite a few places using std::map. Shall we change all of them?

**Comment by mbarbar** [comment]
Good catch. I had changed `Dense*` but not `std::*`.

Currently `SVFSet` and `Map` are set to `std::map` and `std::set`. Their unordered counterparts are faster. Should I switch to those?

Another issue... what if someone needs an ordered set? For example, in `BVDataPTAImpl::dumpAllPts`, the `NodeSet` is used to order the PAG nodes before dumping (the PAG iterator was on a `DenseMap` before, if my memory serves me well). Should we have...

`SVFSet = std::set or std::unordered_set`
`SVFMap = std::map or std::unordered_map`
`SVFOrderedMap = std::map`
`SVFOrderedSet = std::set`

So then, `SVFSet` and `SVFMap` can be set to anything (ordered or unordered: the user can't rely on that knowledge except that they are maps or sets), but `SVFOrderedSet` and `SVFOrderedMap` will *always* be ordered, no matter what implementation they use (LLVM, Boost, STL, ...).

Finally, should we drop the `SVF` prefix and just call them `Map` and `Set`? May be simpler. They are under the `SVF` namespace afterall... `SVF::SVFMap` vs `SVF::Map`.

**Comment by yuleisui** [comment]
> Good catch. I had changed `Dense*` but not `std::*`.
> 
> Currently `SVFSet` and `Map` are set to `std::map` and `std::set`. Their unordered counterparts are faster. Should I switch to those?
> 
> Another issue... what if someone needs an ordered set? For example, in `BVDataPTAImpl::dumpAllPts`, the `NodeSet` is used to order the PAG nodes before dumping (the PAG iterator was on a `DenseMap` before, if my memory serves me well). Should we have...
> 
> `SVFSet = std::set or std::unordered_set`
> `SVFMap = std::map or std::unordered_map`
> `SVFOrderedMap = std::map`
> `SVFOrderedSet = std::set`
> 
Yes, this is a good idea!

> So then, `SVFSet` and `SVFMap` can be set to anything (ordered or unordered: the user can't rely on that knowledge except that they are maps or sets), but `SVFOrderedSet` and `SVFOrderedMap` will _always_ be ordered, no matter what implementation they use (LLVM, Boost, STL, ...).
> 
> Finally, should we drop the `SVF` prefix and just call them `Map` and `Set`? May be simpler. They are under the `SVF` namespace afterall... `SVF::SVFMap` vs `SVF::Map`.
> 
Sounds good to me.
> Edit: let me fix the CI failure.

**Comment by mbarbar** [comment]
Fixed and added, the commit messages should explain rationale.

Still, I am unsure of the `operator==` I implemented for `LocationSet`; there is some commented code in `operator<` and I do not know why it is commented.

**Comment by yuleisui** [comment]
This change is significant. Just in case, can you test a few bcs to see whether this pull request will produce the same points-to results as the old one?

**Comment by mbarbar** [comment]
For `-fspta` and `-ander`, the following produce the same average top level points-to set sizes as the old implementation (tested against a commit just before the `PTData` refactor):

du.bc
i3.bc
mruby.bcc
dpkg.bcc
ninja.bcc
janet.bc
tmux.bc
nvim.bc
bake.bc

**Comment by yuleisui** [comment]
Any benchmarks/programs produce different `average top-level points-to set sizes`, if not I will merge it now.

**Comment by mbarbar** [comment]
I ran tests on those only (nvim and bake were the problematic ones from the renaming perspective we discussed) and all them were successful. If you'd like me to run on some larger programs, I can but it will take a while to complete.

**Comment by yuleisui** [comment]
Yes, please run some larger ones.

**Comment by mbarbar** [comment]
Sure. I'm running on bash, mutt, and hyriseConsole now.

**Comment by mbarbar** [comment]
Tested against bash, mutt, and hyriseConsole and all is good. I also changed some unordered maps to ordered maps (which object creation may rely on the iteration of) and the issues with nvim and bake appear to be fixed.

**Comment by yuleisui** [comment]
Very Good! 
For DDA, can we change in order map or set to ordered ones because dynamic recursive  traversal in DDA analyses may require deterministic order. I think you can still keep the hash functions you have written, just in case if we are interested in using unordered set later.

**Comment by mbarbar** [comment]
Done. Went through all instances of `Set` and `Map` in `include/DDA` and `lib/DDA` (except I barely touched Stat) and changed them to ordered. There are some instances of types defined outside DDA which I didn't change, like `typedef PAG::CallSiteSet CallSiteSet;`. I also didn't change two variables, `ander_vfns` and `dda_vfns`.

Not sure about testing. Maybe in the future DDA tests in Test-Suite could make the process quicker.

**Pull request merged**